lib/pbench/test/unit/server/test_user_create.py

```diff
@@ -54,74 +61,125 @@ def test_args(server_config, pytestconfig):
-                TestCreateUser.FIRST_NAME_TEXT,
-                TestCreateUser.LAST_NAME_SWITCH,
-                TestCreateUser.LAST_NAME_TEXT,
-                TestCreateUser.ROLE_SWITCH,
-                "ADMIN",
-            ],
-            input=f"{TestCreateUser.PSWD_TEXT}\n",
-        )
-        assert result.exit_code == 0
-        assert (
-            result.stdout
-            == f"{TestCreateUser.PSWD_PROMPT}\n" + "Admin user test_user registered\n"
-        )
-        assert not result.stderr_bytes
+    def test_valid_user_delete(client, server_config, pytestconfig):
+        with client:
+            runner = CliRunner(mix_stderr=False)
+            TestUserManagement.test_args(server_config, pytestconfig)
+
+            result = runner.invoke(user_delete, args=[TestUserManagement.USER_TEXT,])
+            assert result.exit_code == 0
+            assert result.stdout == "User test_user deleted\n"
+            assert not result.stderr_bytes
 
     @staticmethod
     @responses.activate
-    def test_user_creation_with_invalid_role(server_config, pytestconfig):
-        runner = CliRunner(mix_stderr=False)
-        result = runner.invoke(
-            user_create.main,
-            args=[
-                TestCreateUser.USER_SWITCH,
-                TestCreateUser.USER_TEXT,
-                TestCreateUser.EMAIL_SWITCH,
-                TestCreateUser.EMAIL_TEXT,
-                TestCreateUser.FIRST_NAME_SWITCH,
-                TestCreateUser.FIRST_NAME_TEXT,
-                TestCreateUser.LAST_NAME_SWITCH,
-                TestCreateUser.LAST_NAME_TEXT,
-                TestCreateUser.ROLE_SWITCH,
-                "ADMN",
-            ],
-            input=f"{TestCreateUser.PSWD_TEXT}\n",
-        )
-        assert result.exit_code == 2
+    def test_valid_users_list(client, server_config, pytestconfig):
+        with client:
+            runner = CliRunner(mix_stderr=False)
+            TestUserManagement.test_args(server_config, pytestconfig)
+
+            result = runner.invoke(user_list,)
+            assert result.exit_code == 0
+            assert result.stdout == "{0:15}\t{1:15}\t{2:15}\t{3:15}\t{4:20}".format(
+                "Username", "First Name", "Last Name", "Registered On", "Email"
+            ) + "\n{0:15}\t{1:15}\t{2:15}\t{3:15}\t{4:20}\n".format(
+                "test_user", "First", "Last", "2021-04-16", "test@domain.com"
+            )
+            assert not result.stderr_bytes
+
+    @staticmethod
+    @responses.activate
+    def test_valid_users_update(client, server_config, pytestconfig):
+        with client:
+            runner = CliRunner(mix_stderr=False)
+            TestUserManagement.test_args(server_config, pytestconfig)
+
+            result = runner.invoke(
+                user_update,
+                args=[
+                    TestUserManagement.USER_TEXT,
+                    TestUserManagement.EMAIL_SWITCH,
+                    "new_test@domain.com",
```

Comment:
Should there be testing of options other than just username and email?

Reply:
I can add more tests for other options, but if the test for updating email worked that means other options would definitely work, because the behavior is the same

Reply:
The ideal is to have enough unit tests that they collectively drive execution down _every_ path through the code, and they form a record of what we think the behavior should be.  Granted, from the inside, we know that all the options are handled the same way and that there _shouldn't_ be any reason why, if one works, they won't all work, but that assumption could be violated in the future when changes are made to the code, and so we want unit tests to explicitly test __everything__.

Now, as a practical matter, writing and maintaining unit tests is a pain, and it sinks a lot of time and effort without producing any direct value, but having good and comprehensive unit testing in place creates tremendous confidence in our development and allows us to be much more agile in terms of making changes in the future, so it's a totally worthwhile investment in the long run.  (And, unit tests are typically easier to write while you're working on the code that they are intended to support.)